metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java

```diff
@@ -763,17 +747,227 @@ public void testValidateBadManualPartitionAssignments() throws Exception {
+        ListPartitionReassignmentsResponseData currentReassigning =
+            new ListPartitionReassignmentsResponseData().setErrorMessage(null).
+                setTopics(Arrays.asList(new OngoingTopicReassignment().
+                    setName("bar").setPartitions(Arrays.asList(
+                    new OngoingPartitionReassignment().setPartitionIndex(0).
+                        setRemovingReplicas(Collections.emptyList()).
+                        setAddingReplicas(Arrays.asList(0, 1)).
+                        setReplicas(Arrays.asList(1, 2, 3, 4, 0))))));
+        assertEquals(currentReassigning, replication.listPartitionReassignments(null));
+        assertEquals(NONE_REASSIGNING, replication.listPartitionReassignments(Arrays.asList(
+            new ListPartitionReassignmentsTopics().setName("foo").
+                setPartitionIndexes(Arrays.asList(0, 1, 2)))));
+        assertEquals(currentReassigning, replication.listPartitionReassignments(Arrays.asList(
+            new ListPartitionReassignmentsTopics().setName("bar").
+                setPartitionIndexes(Arrays.asList(0, 1, 2)))));
+        ControllerResult<AlterPartitionReassignmentsResponseData> cancelResult =
+            replication.alterPartitionReassignments(
+                new AlterPartitionReassignmentsRequestData().setTopics(Arrays.asList(
+                    new ReassignableTopic().setName("foo").setPartitions(Arrays.asList(
+                        new ReassignablePartition().setPartitionIndex(0).
+                            setReplicas(null))),
+                    new ReassignableTopic().setName("bar").setPartitions(Arrays.asList(
+                        new ReassignablePartition().setPartitionIndex(0).
+                            setReplicas(null))))));
+        assertEquals(ControllerResult.atomicOf(Collections.singletonList(new ApiMessageAndVersion(
+                new PartitionChangeRecord().setTopicId(barId).
+                    setPartitionId(0).
+                    setLeader(4).
+                    setReplicas(Arrays.asList(2, 3, 4)).
+                    setRemovingReplicas(null).
+                    setAddingReplicas(Collections.emptyList()), (short) 0)),
+            new AlterPartitionReassignmentsResponseData().setErrorMessage(null).setResponses(Arrays.asList(
+                new ReassignableTopicResponse().setName("foo").setPartitions(Arrays.asList(
+                    new ReassignablePartitionResponse().setPartitionIndex(0).
+                        setErrorCode(NO_REASSIGNMENT_IN_PROGRESS.code()).setErrorMessage(null))),
+                new ReassignableTopicResponse().setName("bar").setPartitions(Arrays.asList(
+                    new ReassignablePartitionResponse().setPartitionIndex(0).
+                        setErrorMessage(null)))))),
+            cancelResult);
+        ControllerResult<AlterIsrResponseData> alterIsrResult = replication.alterIsr(
```

Comment:
Is the alterIsr meant to be tested before cancellation? If so, probably move the code up to make it clear.